File: fix_jsx_closing_tags.py
#!/usr/bin/env python3
"""
Script to fix JSX closing tag errors.
"""

import os
import re
import glob

def fix_jsx_closing_tags(file_path):
    """Fix JSX closing tag errors."""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        original_content = content
        
        # Fix common JSX issues
        fixes = [
            # Fix missing closing div tags
            (r'<div([^>]*)>\s*$', r'<div\1></div>'),
            # Fix missing closing fragment tags
            (r'<>\s*$', r'<></>'),
            # Fix missing closing section tags
            (r'<section([^>]*)>\s*$', r'<section\1></section>'),
            # Fix missing closing main tags
            (r'<main([^>]*)>\s*$', r'<main\1></main>'),
            # Fix missing closing article tags
            (r'<article([^>]*)>\s*$', r'<article\1></article>'),
            # Fix missing closing header tags
            (r'<header([^>]*)>\s*$', r'<header\1></header>'),
            # Fix missing closing footer tags
            (r'<footer([^>]*)>\s*$', r'<footer\1></footer>'),
            # Fix missing closing nav tags
            (r'<nav([^>]*)>\s*$', r'<nav\1></nav>'),
            # Fix missing closing aside tags
            (r'<aside([^>]*)>\s*$', r'<aside\1></aside>'),
        ]
        
        # Apply fixes
        for pattern, replacement in fixes:
            content = re.sub(pattern, replacement, content, flags=re.MULTILINE)
        
        # Fix specific patterns that might be causing issues
        # Fix missing commas in object literals
        content = re.sub(r'(\w+):\s*(\w+)\s*$', r'\1: \2,', content)
        
        # Fix missing closing parentheses
        content = re.sub(r'(\w+)\s*$', r'\1)', content)
        
        # Write back if changes were made
        if content != original_content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            print(f"Fixed JSX closing tags in: {file_path}")
            return True
        
    except Exception as e:
        print(f"Error processing {file_path}: {e}")
        return False

def main():
    """Main function to fix JSX closing tags."""
    patterns = [
        'app/**/*.tsx',
        'app/**/*.ts',
        'components/**/*.tsx',
        'components/**/*.ts'
    ]
    
    files_processed = 0
    files_fixed = 0
    
    for pattern in patterns:
        for file_path in glob.glob(pattern, recursive=True):
            # Skip .original files
            if '.original' in file_path:
                continue
                
            files_processed += 1
            if fix_jsx_closing_tags(file_path):
                files_fixed += 1
    
    print(f"\nProcessed {files_processed} files")
    print(f"Fixed JSX closing tags in {files_fixed} files")

if __name__ == "__main__":
    main()

File: fix-jsx-closing-tags.js
#!/usr/bin/env node

import fs from 'fs';
import path from 'path';
import { execSync } from 'child_process';

// Get all TSX files
const tsxFiles = execSync('find . -name "*.tsx" -not -path "./node_modules/*" -not -path "./dist/*" | head -100', { encoding: 'utf8' })
  .trim()
  .split('\n')
  .filter(file => file);

console.log(`Found ${tsxFiles.length} TSX files to check`);

function fixJSXClosingTags(content) {
  let fixed = content;
  
  // Fix missing closing > in JSX tags
  // Pattern: <TagName followed by newline or whitespace without >
  fixed = fixed.replace(/<(\w+)([^>]*?)(?<!\s\/)\s*$/gm, '<$1$2>');
  
  // Fix specific common patterns
  fixed = fixed.replace(/<ErrorBoundary\s*$/gm, '<ErrorBoundary>');
  fixed = fixed.replace(/<HelmetProvider\s*$/gm, '<HelmetProvider>');
  fixed = fixed.replace(/<Router\s*$/gm, '<Router>');
  fixed = fixed.replace(/<Routes\s*$/gm, '<Routes>');
  fixed = fixed.replace(/<Suspense\s*$/gm, '<Suspense>');
  fixed = fixed.replace(/<main\s*$/gm, '<main>');
  fixed = fixed.replace(/<div\s*$/gm, '<div>');
  fixed = fixed.replace(/<section\s*$/gm, '<section>');
  fixed = fixed.replace(/<h1\s*$/gm, '<h1>');
  fixed = fixed.replace(/<h2\s*$/gm, '<h2>');
  fixed = fixed.replace(/<h3\s*$/gm, '<h3>');
  fixed = fixed.replace(/<h4\s*$/gm, '<h4>');
  fixed = fixed.replace(/<h5\s*$/gm, '<h5>');
  fixed = fixed.replace(/<h6\s*$/gm, '<h6>');
  fixed = fixed.replace(/<p\s*$/gm, '<p>');
  fixed = fixed.replace(/<span\s*$/gm, '<span>');
  fixed = fixed.replace(/<button\s*$/gm, '<button>');
  fixed = fixed.replace(/<ul\s*$/gm, '<ul>');
  fixed = fixed.replace(/<li\s*$/gm, '<li>');
  fixed = fixed.replace(/<React\.Fragment\s*$/gm, '<React.Fragment>');
  
  // Fix missing closing > in self-closing tags
  fixed = fixed.replace(/<(\w+)([^>]*?)\s*$/gm, (match, tagName, attributes) => {
    if (attributes.includes('/')) return match; // Already self-closing
    if (tagName.toLowerCase() === 'img' || tagName.toLowerCase() === 'br' || tagName.toLowerCase() === 'hr') {
      return match + '/>'; // Self-closing tags
    }
    return match + '>'; // Regular tags
  });
  
  return fixed;
}

let processedCount = 0;
let errorCount = 0;

for (const filePath of tsxFiles) {
  try {
    if (!fs.existsSync(filePath)) {
      console.log(`File not found: ${filePath}`);
      continue;
    }
    
    const originalContent = fs.readFileSync(filePath, 'utf8');
    let fixedContent = fixJSXClosingTags(originalContent);
    
    // Only write if content changed
    if (fixedContent !== originalContent) {
      fs.writeFileSync(filePath, fixedContent, 'utf8');
      console.log(`Fixed: ${filePath}`);
      processedCount++;
    }
  } catch (error) {
    console.error(`Error processing ${filePath}:`, error.message);
    errorCount++;
  }
}

console.log(`\nProcessed ${processedCount} files, ${errorCount} errors`);